lib/preview-web/src/PreviewWeb.tsx

```diff
@@ -0,0 +1,559 @@
+    // if a subsequent *re-render* is in progress, but that is less likely)
+    // See also the note about cancelling below.
+    const rerenderStory = async () => {
+      // The story has not finished rendered the first time. The loaders are still running
+      // and we will pick up the new args/globals values when renderToDOM is called.
+      if (initialRenderPhase === 'init') {
+        return;
+      }
+
+      // This story context will have the updated values of args and globals
+
+      const rerenderStoryContext = {
+        // NOTE: loaders are not getting run again. So we are just patching
+        // the updated story context over the previous value (that included loader output).
+        // Loaders aren't allowed to touch anything but the `loaded` key but
+        // this means loaders never run again with new values of args/globals
+        ...renderContext.storyContext,
+        ...this.storyStore.getStoryContext(story),
+      };
+      const rerenderRenderContext: RenderContext<TFramework> = {
+        ...renderContext,
+        forceRemount: false,
+        storyContext: rerenderStoryContext,
+        storyFn: () => unboundStoryFn(rerenderStoryContext),
+      };
+
+      try {
+        await this.renderToDOM(rerenderRenderContext, element);
+      } catch (err) {
+        renderContextWithoutStoryContext.showException(err);
+        return;
+      }
+      this.channel.emit(Events.STORY_RENDERED, id);
+    };
+
+    // Start the first render
+    // NOTE: we don't await here because we need to return the "cleanup" function below
+    // right away, so if the user changes story during the first render we can cancel
+    // it without having to first wait for it to finish.
+    initialRender().catch((err) => renderContextWithoutStoryContext.showException(err));
+
+    // Listen to events and re-render story
+    this.channel.on(Events.UPDATE_GLOBALS, rerenderStory);
+    this.channel.on(Events.FORCE_RE_RENDER, rerenderStory);
+    const rerenderStoryIfMatches = async ({ storyId }: { storyId: StoryId }) => {
+      if (storyId === story.id) rerenderStory();
+    };
+    this.channel.on(Events.UPDATE_STORY_ARGS, rerenderStoryIfMatches);
+    this.channel.on(Events.RESET_STORY_ARGS, rerenderStoryIfMatches);
+
+    return () => {
+      // If the story is torn down (either a new story is rendered or the docs page removes it)
+      // we need to consider the fact that the initial render may not be finished
+      // (possibly the loaders or the play function are still running). We use the controller
+      // as a method to abort them, ASAP, but this is not foolproof as we cannot control what
+      // happens inside the user's code. Still, we do render the new story right away.
+      // Alternatively, we could make this function async and await the teardown before rendering
+      // the new story. This might be a bit complicated for docs however.
+      controller.abort();
+      this.storyStore.cleanupStory(story);
```

Comment:
Just a thought: I'm wondering if it makes sense to expose an event or a cleanup utility for end users when this happens. As we keep progressing to having more complex components and their interactions on Storybook, there will likely be a need for cleaning up/restoring mocks and such. Maybe it could be a cleanup function defined in the story which is taken into account in the cleanupStory method.

I'm not really sure if this would be necessary, as likely the cleanup (like unregistering handlers) would be done in a component's unmount process, but maybe people will want to do some stuff at the level of the `play` function and want to clean that up as well?

Does this make sense?

Reply:
Definitely makes sense as something we could add down the track!

Reply:
Great 👌